rust/kernel/regmap.rs

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::error::{ptr_err_check, Error, KernelResult};
+use crate::platform_driver::{Device, PlatformDevice};
+use crate::{bindings, c_types};
+use core::mem::MaybeUninit;
+
+/// Abstraction wrapping a kernel `struct regmap`.
+///
+/// # Invariants
+///
+/// regmap locking is never disabled.
+pub struct Regmap(*mut bindings::regmap);
+
+// SAFETY: we access the underlying `struct regmap` only through its helper functions,
+// and we never disable locking by the type invariant, so:
+// - `struct regmap *` can be safely sent between threads (Send)
+// - `struct regmap`'s helper functions can be safely called from any thread (Sync)
+unsafe impl Send for Regmap {}
+unsafe impl Sync for Regmap {}
+
+impl Regmap {
+    pub fn write(&self, reg: c_types::c_uint, val: c_types::c_uint) -> KernelResult {
+        // SAFETY: FFI call.
+        // OK to coerce a shared reference to a mutable pointer, as
+        // we guarantee that a `struct regmap`'s `write` is fully synchronized.
+        let res = unsafe { bindings::regmap_write(self.0, reg, val) };
+        if res != 0 {
+            return Err(Error::from_kernel_errno(res));
+        }
+        Ok(())
+    }
+
+    pub fn read(&self, reg: c_types::c_uint) -> KernelResult<c_types::c_uint> {
+        let mut val = MaybeUninit::<c_types::c_uint>::uninit();
+        // SAFETY: FFI call.
+        // OK to coerce a shared reference to a mutable pointer, as
+        // we guarantee that a `struct regmap`'s `read` is fully synchronized.
+        // OK to pass a pointer to an uninitialized `u32`, this is part of
+        // `struct regmap`'s `read` API.
+        let res = unsafe { bindings::regmap_read(self.0, reg, val.as_mut_ptr()) };
+        if res != 0 {
+            return Err(Error::from_kernel_errno(res));
+        }
+        // SAFETY: if `res` is zero, `val` is guaranteed initialized by the
+        // call to `regmap_read`.
+        Ok(unsafe { val.assume_init() })
+    }
+
+    #[cfg(CONFIG_REGMAP_MMIO)]
+    pub fn init_mmio_platform_resource(
+        pdev: &mut PlatformDevice,
+        index: u32,
+        cfg: &RegmapConfig,
+    ) -> KernelResult<Self> {
+        let iomem = devm_platform_ioremap_resource(pdev, index)?;
+        devm_regmap_init_mmio(pdev, iomem, cfg)
+    }
+}
+
+#[derive(Default)]
+pub struct RegmapConfig {
+    reg_bits: i32,
+    val_bits: i32,
+    reg_stride: Option<i32>,
+    max_register: Option<u32>,
+}
+
+impl RegmapConfig {
+    pub fn new(reg_bits: u32, val_bits: u32) -> RegmapConfig {
+        RegmapConfig {
+            reg_bits: reg_bits as i32,
+            val_bits: val_bits as i32,
+            ..Default::default()
+        }
+    }
+
+    pub fn reg_stride(mut self, reg_stride: u32) -> Self {
+        self.reg_stride.replace(reg_stride as i32);
```

Comment:
This could simply overwrite since you don't need the old value.

Also, why do we need the parameter to be `u32`? Is it to ensure it is positive?

```suggestion
        self.reg_stride = Some(reg_stride as i32);
```

Reply:
Yes, I can see no possibly way that `stride` could be negative. So why allow people to do this in Rust?

Reply:
Then, do we need to check for overflow?

Reply:
I could use a `u16` here, which avoids the overflow issue. Register strides cannot sensibly take values greater than a few multiples of the machine word length. Or, just stick with the kernel's `i32`, so when the driver writer messes up and passes a negative value, `regmap_init()` will just naturally error out. What's your opinion?

Reply:
In general, use the most stringent choice you can.

While in C traditionally one uses `int` as a sensible & fast default when the values are expected to fit in any architecture's `int`; in Rust we need to be explicit with the bit length anyway (except for `usize` etc., but that is a different use case).

Being stringent has other benefits. For instance, if you go for a type that always fit, you can then actually write:

    self.reg_stride = Some(reg_stride.into());

Which ensures the program will fail to compile if someone changes the type back to something that does not always fit.

And callers are fine, because non-suffixed literals will still compile:

    ...with_reg_stride(42)

So If a `u8` or `u16` is enough, then definitely go with that!